ftdc/sys/sys.go

```diff
@@ -0,0 +1,100 @@
+// Package sys provides functionality for gathering system metrics in an FTDC compliant API.
+package sys
+
+import (
+	"os"
+	"time"
+
+	"github.com/prometheus/procfs"
+)
+
+// On linux, getting the page size is a system call. Cache the page size for the entirety of the
+// progarm lifetime. As opposed to calling it each time we wish to compute the resident memory a
+// program is using.
+var (
+	osPageSize                    int
+	machineBootTimeSecsSinceEpoch float64
+)
+
+func init() {
+	osPageSize = os.Getpagesize()
+
+	machine, err := procfs.NewDefaultFS()
+	if err != nil {
+		return
+	}
+
+	machineStats, err := machine.Stat()
+	if err != nil {
+		return
+	}
+
+	machineBootTimeSecsSinceEpoch = float64(machineStats.BootTime)
+}
+
+// UsageStatser can be used to get system metrics for a process.
+type UsageStatser struct {
+	proc procfs.Proc
+}
+
+// NewSelfSysUsageStatser will return a `SysUsageStatser` for the current process.
+func NewSelfSysUsageStatser() (*UsageStatser, error) {
+	process, err := procfs.Self()
+	if err != nil {
+		return nil, err
+	}
+
+	return &UsageStatser{process}, nil
+}
+
+// NewPidSysUsageStatser will return a `SysUsageStatser` for the given process id.
+func NewPidSysUsageStatser(pid int) (*UsageStatser, error) {
+	process, err := procfs.NewProc(pid)
+	if err != nil {
+		return nil, err
+	}
+
+	return &UsageStatser{process}, nil
+}
+
+type stats struct {
+	UserCPUSecs     float64
+	SystemCPUSecs   float64
+	ElapsedTimeSecs float64
+	VssMB           float64
+	RssMB           float64
+}
+
+// Stats returns Stats.
+func (sys *UsageStatser) Stats() any {
```

Comment:
what does Stats() return on a mac or possibly a windows machine?

Reply:
I expect the constructor to fail on both -- therefore FTDC would never call `Stats`. But if the constructor succeeds, I'd expect the `Stat` call on line 80 to return an error and we'd just get 0 values out.